### PR TITLE
add append_version feature into StaticResource url resolver

### PR DESCRIPTION
**Pull request by sky-code**
## Are there changes in behavior for the user?
Fully backward compatible feature (has no impact on exists code)

## Related issue number
Please goto #2157 for description and discussion

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `changes` folder
  * name it `<issue_id>.<type>` for example (588.bug)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by asvetlov** [comment]
Please mention new parameter and behavior in docs

**Comment by sky-code** [comment]
I have added documentation for `url_for` but not for `url` because it's deprecated

**Comment by asvetlov** [comment]
The PR only generates new version tag by `url_for()` call.

Should it return redirection response on version mismatch?

Should we move `append_version` parameter to constructor? 

It might simplify the usage: just add static files handler with parameter set (`app.router.add_static('/prefix', path_to_static, append_version=True)`), no need to pass `append_version=True` to every `url_for()` call.

**Comment by sky-code** [comment]
Should it return redirection response on version mismatch?
Not necessarily, no at this time. If someone asks to make it and provide good explanation of advantages then it will be added

Should we move append_version parameter to constructor?
maybe, set default value in constructor than override it in each call if necessary

File: aiohttp/web_urldispatcher.py [hunk 1]
```diff
@@ -450,17 +453,51 @@ def __init__(self, prefix, directory, *, name=None,
...
                                               expect_handler=expect_handler)}
 
-    def url(self, *, filename, query=None):
-        return str(self.url_for(filename=filename).with_query(query))
+    def url(self, *, filename, append_version=False, query=None):
+        # is there with_query need to be used?
+        # with_query remove previous query options
+        url = self.url_for(filename=filename, append_version=append_version)
+        if query is not None:
+            return str(url.with_query(query))
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Should be `update_query`

File: aiohttp/web_urldispatcher.py [hunk 2]
```diff
@@ -450,17 +453,51 @@ def __init__(self, prefix, directory, *, name=None,
...
+                # TODO cache file content
+                # with file watcher for cache invalidation
+                with open(str(filepath), mode='rb') as f:
+                    file_bytes = f.read()
+                h = self.get_file_hash(file_bytes)
+                url = url.with_query({self.VERSION_KEY: h})
+                return url
+        return url
+
+    def get_file_hash(self, byte_array):
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Add `_` prefix to method name: it is not a part of public API

File: aiohttp/web_urldispatcher.py [hunk 3]
```diff
@@ -450,17 +453,51 @@ def __init__(self, prefix, directory, *, name=None,
...
+                    filename = filename[1:]
+                filepath = self._directory.joinpath(filename).resolve()
+                if not self._follow_symlinks:
+                    filepath.relative_to(self._directory)
+            except (ValueError, FileNotFoundError) as error:
+                # relatively safe
+                return url
+            except Exception as error:
+                # perm error or other kind!
+                # TODO log error here?
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Not sure if we should catch the exception.
url is useless: next GET will raise an error.
Moreover next `open` call might fail too.

File: aiohttp/web_urldispatcher.py [hunk 4]
```diff
@@ -450,17 +453,51 @@ def __init__(self, prefix, directory, *, name=None,
...
-        return URL(url)
+        url = URL(url)
+        if append_version is True:
+            try:
+                if filename.startswith('/'):
+                    filename = filename[1:]
+                filepath = self._directory.joinpath(filename).resolve()
+                if not self._follow_symlinks:
+                    filepath.relative_to(self._directory)
+            except (ValueError, FileNotFoundError) as error:
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: What case could raise `ValueError`?

Comment (sky-code) [reply aiohttp/web_urldispatcher.py]: I don't know, this is copied code from `_handle` method, anyway i remove it

File: aiohttp/web_urldispatcher.py (hunk 3 above)
Comment (sky-code) [reply aiohttp/web_urldispatcher.py]: removed

File: aiohttp/web_urldispatcher.py (hunk 2 above)
Comment (sky-code) [reply aiohttp/web_urldispatcher.py]: done

File: aiohttp/web_urldispatcher.py (hunk 1 above)
Comment (sky-code) [reply aiohttp/web_urldispatcher.py]: done

File: aiohttp/web_urldispatcher.py [hunk 5]
```diff
@@ -443,24 +447,57 @@ def __init__(self, prefix, directory, *, name=None,
...
-    def url(self, *, filename, query=None):
-        return str(self.url_for(filename=filename).with_query(query))
-
-    def url_for(self, *, filename):
+    def url(self, *, filename, append_version=None, query=None):
+        # is there with_query need to be used?
+        # with_query remove previous query options
+        url = self.url_for(filename=filename, append_version=append_version)
+        if query is not None:
+            return str(url.update_query(query))
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Please add a test for this case

File: aiohttp/web_urldispatcher.py [hunk 6]
```diff
@@ -443,24 +447,57 @@ def __init__(self, prefix, directory, *, name=None,
...
             filename = str(filename)
         while filename.startswith('/'):
             filename = filename[1:]
         filename = '/' + filename
         url = self._prefix + URL(filename).raw_path
-        return URL(url)
+        url = URL(url)
+        if append_version is True:
+            try:
+                if filename.startswith('/'):
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py:476]: Please add a test if `filename` doesn't start with `/`

File: aiohttp/web_urldispatcher.py [hunk 7]
```diff
@@ -443,24 +447,57 @@ def __init__(self, prefix, directory, *, name=None,
...
+                    filename = filename[1:]
+                filepath = self._directory.joinpath(filename).resolve()
+                if not self._follow_symlinks:
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py:479]: Test for `_follow_symlink` needed

File: aiohttp/web_urldispatcher.py (hunk 5 above)
Comment (sky-code) [reply aiohttp/web_urldispatcher.py]: Done

File: aiohttp/web_urldispatcher.py (hunk 6 above)
Comment (sky-code) [reply aiohttp/web_urldispatcher.py:476]: Done

File: aiohttp/web_urldispatcher.py (hunk 7 above)
Comment (sky-code) [reply aiohttp/web_urldispatcher.py:479]: Done

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**